rune/libenclave/internal/runtime/pal/skeleton/sgxutils.c

```diff
@@ -49,44 +49,34 @@ static inline uint64_t get_xcr0()
 	return xgetbv(0);
 }
 
-static bool try_get_xcr0(uint64_t *value)
+static uint64_t try_get_xcr0()
 {
 	int cpu_info[4] = {0, 0, 0, 0};
 
-	*value = SGX_XFRM_LEGACY;
-
-	// check if xgetbv instruction is supported
+	// check if xgetbv instruction is supported.
 	__cpuid(cpu_info, 1);
-	// ecx[27:26] indicate whether support xsave/xrstor, and whether enable xgetbv, xsetbv
-	if (!(cpu_info[2] & (1<<XSAVE_SHIFT)) || !(cpu_info[2] & (1<<OSXSAVE_SHIFT)))
-		return false;
-
-	*value = get_xcr0();
+	// ecx[27:26] indicate whether support xsave/xrstor, and whether enable xgetbv, xsetbv.
+	if (!(cpu_info[2] & (1<<OSXSAVE_SHIFT)))
+		return SGX_XFRM_LEGACY;
 
-	// check if xsavec is supported
-	// Assume that XSAVEC is always supported if XSAVE is supported
+	// check if xsavec is supported.
+	// Assume that XSAVEC is always supported if XSAVE is supported.
+	// if XSAVE is supported, while XSAVEC is not supported,
+	// set xfrm to legacy, because XSAVEC cannot be executed within enclave.
 	cpu_info[0] = cpu_info[1] = cpu_info[2] = cpu_info[3] = 0;
 	__cpuidex(cpu_info, 0xD, 1);
 	if (!(cpu_info[0] & (1<<XSAVEC_SHIFT)))
-		return false;
+		return SGX_XFRM_LEGACY;
 
-	return true;
+	// If x-feature is supported and enabled by OS, we need make sure it is also supported in enclave.
+	cpu_info[0] = cpu_info[1] = cpu_info[2] = cpu_info[3] = 0;
```

Comment:
This line is not necessary because this cpuid leaf function will overwrite existing values.